src/python/pants/base/specs.py

```diff
@@ -223,12 +240,14 @@ class SpecsMatcher:
   the hash of the class instance in its key, and results in a very large key when used with `Specs`
   directly).
   """
-  tags: Tuple
-  exclude_patterns: Tuple
+  tags: Tuple[str, ...]
+  exclude_patterns: Tuple[str, ...]
 
-  def __init__(self, tags: Optional[Tuple] = None, exclude_patterns: Tuple = ()) -> None:
+  def __init__(
+    self, tags: Optional[Iterable[str]] = None, exclude_patterns: Optional[Iterable[str]] = None,
```

Comment:
I changed `exclude_patterns` to be `Optional` so that it mirrors `tags` and avoids having to add `or ()` to some downstream code. It didn't seem like there was a strong reason why this wasn't done in the first place.